public/app/core/components/connectWithCleanUp.tsx

```diff
@@ -0,0 +1,36 @@
+import { MapStateToPropsParam, MapDispatchToPropsParam, connect, useDispatch } from 'react-redux';
+import { StateSelector, cleanUpAction } from '../actions/cleanUp';
+import React, { ComponentType, FunctionComponent, useEffect } from 'react';
+import hoistNonReactStatics from 'hoist-non-react-statics';
+
+export const connectWithCleanUp = <
+  TStateProps extends {} = {},
+  TDispatchProps = {},
+  TOwnProps = {},
+  State = {},
+  TSelector extends object = {}
+>(
+  mapStateToProps: MapStateToPropsParam<TStateProps, TOwnProps, State>,
+  mapDispatchToProps: MapDispatchToPropsParam<TDispatchProps, TOwnProps>,
+  stateSelector: StateSelector<TSelector>
+) => (Component: ComponentType<any>) => {
+  const ConnectedComponent = connect(
+    mapStateToProps,
+    mapDispatchToProps
+  )(Component);
+
+  const ConnectedComponentWithCleanUp: FunctionComponent = props => {
+    const dispatch = useDispatch();
+    useEffect(() => {
+      return () => dispatch(cleanUpAction({ stateSelector }));
+    }, []);
+    // @ts-ignore
+    return <ConnectedComponent {...props} />;
+  };
```

Comment:
since we are wrapping the page component in a functional component, won't the lack of React.memo() here make it re-render on every state change (removing the PureComponent props changed check)

Reply:
Yes of course, very good catch! Will update

Reply:
I dug a bit deeper on this and it seems like the `connect` HOC is already using `memo` so the `ConnectWithCleanUp` functional component is not re rendered as far as I can see from the React Dev Tools profiling. I could use some eyes on my assumption here so I'm looking at this wrong.

The top TeamList in the pics below is the `ConnectWithCleanUp`. Grey means that it's not re rendered when I dispatch this action:
```js
{
  type: 'LOAD_TEAMS',
  payload: [
    {
      id: 1,
      orgId: 1,
      name: 'kalles',
      email: '',
      avatarUrl: '/avatar/740780ec0160a02989bfb672220f1307',
      memberCount: 1,
      permission: 0
    },
{
      id: 2,
      orgId: 1,
      name: 'mias',
      email: '',
      avatarUrl: '/avatar/740780ec0160a02989bfb672220f1307',
      memberCount: 4,
      permission: 0
    }
  ]
}
```
**Without memo**
![image](https://user-images.githubusercontent.com/562238/66190182-c0bd8800-e68b-11e9-9a77-c93700f34f9b.png)

**With memo**
![image](https://user-images.githubusercontent.com/562238/66190333-1003b880-e68c-11e9-810c-0dddc5991fda.png)

Reply:
ah, great. that is good to know :)